### PR TITLE
secp256k1: Optimize k splitting with mod n scalar.

### PR DESCRIPTION
**Pull request by davecgh**
~**This requires #2887**.~

This optimizes the scalar decomposition code by rewriting it to make use of the highly-efficient zero-allocation `ModNScalar` type along with `math/bits` (which provides hardware acceleration on most platforms) instead of `big.Ints`.

The net effect is that the decomposition is significantly faster, allocation free, and constant time.

Since this is touching code that is critical to consensus, and the math involved might not be immediately familiar to reviewers, it also adds a bunch of detailed comments to better describe the endomorphism and how it is used in the scalar multiplication in addition to fully describing the scalar decomposition process and derivation.

Finally, it adds logic to derive and print all of the constants the reworked code makes of to the precomputation code that runs with `go generate`.

The following is a high level overview of the changes:

- Adds a benchmark for the scalar decomposition
- Adds derivation code for all precomputed constants
- Reworks associated tests to make the testing methodology less reliant on the specific implementation details and to test edge cases and to additionally assert the following properties:
  - The decomposed scalars must result in the original scalar via the equation `k = k1 + k2*λ (mod n)`
  - Both decomposed scalars must have small magnitude as determined by having a maximum bit length of one more than half of the bit size of the underlying field
- Updates the random k splitting tests to ensure new random values are tested each run
- Rewrites `splitK` to use the `ModNScalar` type instead of `big.Ints`:
  - Allocation free
  - Constant time
  - Provides hardware acceleration on most platforms
  - Includes detailed comments describing the scalar decomposition process and derivation
- Updates endomorphism parameter constant definitions to be `ModNScalars` instead of `big.Int`
      - Moves the convenience func `hexToModNScalar` to the main package
- Adds new `endoZ1` and `endoZ2` constants for the rounded multiplication used by the new decomposition code
  - Adds logic to derive and print the new endomorphism constants with `go generate`
- Updates the scalar multiplication to account for the new semantics
- Adds detailed comments to scalar multiplication
- Tightens negation magnitudes in `addZ1EqualsZ2` and remove no longer needed normalization
- Ensures the calculation when recovering compact signatures uses normalized points
 

The following benchmarks show a before and after comparison of scalar decomposition as well as how it that translates to scalar multiplication and signature verfication:

```
name         old time/op    new time/op     delta
---------------------------------------------------------------------
SplitK       1.61µs ±32%    0.89µs ± 2%     -44.69% (p=0.000 n=50+47)
ScalarMult   125µs ± 1%     115µs ± 1%      -7.82%  (p=0.000 n=43+46)
SigVerify    161µs ±25%     160µs ±19%      -0.53%  (p=0.001 n=50+50)

name         old allocs/op  new allocs/op   delta
-----------------------------------------------------------------------
SplitK       10.0 ± 0%       0.0            -100.00%  (p=0.000 n=50+50)
ScalarMult   11.0 ± 0%       0.0            -100.00%  (p=0.000 n=50+50)
SigVerify    28.0 ± 0%      16.0 ± 0%       -42.86%   (p=0.000 n=50+50)
```

While it only saves about 1 µs per signature verification in the benchmarking scenario, the primary win is the reduction in the number of allocations per signature verification which has a much more significant impact when verifying large numbers of signatures back to back, such as when processing new blocks, and especially during the initial chain sync process.

### CODE REVIEW

**Review by matheusd — APPROVED** [review]
Woop Woop!

It's invaluable having all the rationale for the calculations inline with the code. Makes reviewing the correctness of the code **much** easier.

I've marked some places inline where I've manually checked the assumptions. I also had `GECC` opened here to follow the explanations, but the inline, step-by-step derivations are much easier for me to follow.

All in all, excellent work as usual!

File: dcrec/secp256k1/curve.go [hunk 1]
```diff
@@ -597,47 +629,282 @@ func DoubleNonConst(p, result *JacobianPoint) {
...
+	// algorithm and are precomputed since both n and λ are fixed values for the
+	// secp256k1 curve.  See genprecomps.go for derivation details.
+	//
+	// Next, consider k as a vector <k, 0> in ℚ⨯ℚ and by linear algebra write:
+	//
+	// <k, 0> = g1*v1 + g2*v2, where g1, g2 ∈ ℚ
+	//
+	// Note that, per above, the components of vector v1 are a1 and b1 while the
+	// components of vector v2 are a2 and b2.  Given the vectors v1 and v2 were
+	// generated such that a1*b2 - a2*b1 = n, solving the equation for g1 and g2
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve.go:795]: Recall that we're working in the rationals here ("`where g1, g2 ∈ ℚ`") so the (mod n) bits don't apply.  It's `n` because the vectors are generated via the Extended Euclidean Algorithm (EEA) which generates a sequence of equations `s[i]*n + t[i]*λ = r[i]` (because `n` and `λ` are the two inputs).

The vectors are chosen from three consecutive iterations as soon as the remainder becomes less than `sqrt(n)` as `a1 = r[i+1], b1 = -t[i+1]` and either `a2 = r[i], b2 = -t[i]` or `a2 = r[i+2], b2 = -t[i+2]` depending on which has the smaller Euclidean norm.  

Since the EEA works by consecutively dividing the previous remainders to obtain an integer quotient and a new remainder, it follows that `a1*b2 - a2*b1` alternates between `n` and `-n` and, as noted above, the iteration on whichever side of the the `i+1` iteration with the smaller Euclidean is chosen which forces the value to be `n`.

**Comment by davecgh** [comment]
Thanks for the thorough review @matheusd.

File: dcrec/secp256k1/curve.go [hunk 2]
```diff
@@ -597,47 +629,282 @@ func DoubleNonConst(p, result *JacobianPoint) {
...
+	//    b1 and b2.
+	//
+	// k2 = c1*(-b1) + c2*(-b2)
+	// k1 = k - k2*λ
+	var k1, k2 ModNScalar
+	c1 := mul512Rsh320Round(k, endoZ1)
+	c2 := mul512Rsh320Round(k, endoZ2)
+	k2.Add2(c1.Mul(endoNegB1), c2.Mul(endoNegB2))
+	k1.Mul2(&k2, endoLambda).Negate().Add(k)
+	return k1, k2
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve.go:910]: FYI: I updated this to shave an extra field negation off by storing the negative version of lambda.

**Review by dajohi — APPROVED** [review]

**Review by rstaudt2 — COMMENTED** [review]
This looks great.  I was not familiar with some of the math/optimizations beforehand but was able to follow with the excellent comments and references.  I also tested by running a full sync on mainnet with `--nocheckpoints` and `--assumevalid=0`.

File: dcrec/secp256k1/curve.go [hunk 3]
```diff
@@ -243,10 +275,10 @@ func addZ1EqualsZ2(p1, p2, result *JacobianPoint) {
...
 	f.Mul2(x2, &b)                         // F = X2*B (mag: 1)
-	x3.Add2(&e, &f).Negate(3).Add(&d)      // X3 = D-E-F (mag: 5)
-	negX3.Set(x3).Negate(5).Normalize()    // negX3 = -X3 (mag: 1)
-	y3.Set(y1).Mul(f.Add(&negE)).Negate(3) // Y3 = -(Y1*(F-E)) (mag: 4)
-	y3.Add(e.Add(&negX3).Mul(&c))          // Y3 = C*(E-X3)+Y3 (mag: 5)
+	x3.Add2(&e, &f).Negate(2).Add(&d)      // X3 = D-E-F (mag: 4)
+	negX3.Set(x3).Negate(4)                // negX3 = -X3 (mag: 5)
+	y3.Set(y1).Mul(f.Add(&negE)).Negate(1) // Y3 = -(Y1*(F-E)) (mag: 2)
+	y3.Add(e.Add(&negX3).Mul(&c))          // Y3 = C*(E-X3)+Y3 (mag: 3)
 	z3.Mul2(z1, &a)                        // Z3 = Z1*A (mag: 1)
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve.go:282]: Having a magnitude of 1 is not the same thing as being normalized.

The comment on `Mul2` calls out:

```Go
// Output Normalized: No
// Output Max Magnitude: 1
func (f *FieldVal) Mul2(val *FieldVal, val2 *FieldVal) *FieldVal {
```

So, it does still need to be normalized.

File: dcrec/secp256k1/curve.go [hunk 4]
```diff
@@ -597,47 +629,285 @@ func DoubleNonConst(p, result *JacobianPoint) {
...
+	//   (2^64 - 1)(2^64 - 1) + 2*(2^64 - 1) = 2^128 - 1
+	// and:
+	//   2^128 - 1 < 2^128.
+	var c2 uint64
+	hi, lo = bits.Mul64(digit1, digit2)
+	lo, c2 = bits.Add64(lo, m, 0)
+	hi, _ = bits.Add64(hi, 0, c2)
+	lo, c2 = bits.Add64(lo, c, 0)
+	hi, _ = bits.Add64(hi, 0, c2)
+	return hi, lo
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve.go:664]: Those are indeed the same.  I originally kept it separate as I thought it made the fact the carry could be dropped a little more obvious, but I don't mind updating it as suggested if you think it makes it clearer.

File: dcrec/secp256k1/curve.go [hunk 5]
```diff
@@ -39,6 +42,33 @@ func hexToFieldVal(s string) *FieldVal {
...
+	}
+	if isNegative {
+		scalar.Negate()
+	}
+	return &scalar
+}
+
 var (
 	// The following constants are used to accelerate scalar point
 	// multiplication through the use of the endomorpishm:
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve.go]: Nice catch.  Fixed in the appropriate commit.

File: dcrec/secp256k1/curve.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rstaudt2) [reply dcrec/secp256k1/curve.go:282]: Ah, got it.  Thanks for clarifying.  The comment below that states `Normalize the resulting field values to a magnitude of 1 as needed` threw me off.  I also noticed that there must not be a test hitting that case since if I comment out the `z3.Normalize()`, all of the tests still pass.

[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve.go:282]: Good point.  That comment does seems to imply otherwise and not what it really means.  I'll give things a pass regarding this to make it more clear.

File: dcrec/secp256k1/curve.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rstaudt2) [reply dcrec/secp256k1/curve.go:664]: Makes sense.  I don't have a strong preference either way.  I mainly brought it up because, at first, I thought the function might be doing something different than the other one since it wasn't calling it.

File: dcrec/secp256k1/curve.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve.go:282]: Oh, I didn't comment about your last bit regarding the tests still passing.

That's because the normalizations are not strictly needed for correctness given the specific way they are used in the existing implementation code because there are enough normalizations elsewhere in the calculations to avoid internal overflows.

However, there is no guarantee on exactly how the exported funcs will be used and their APIs specifically state that:

```Go
// NOTE: The points must be normalized for this function to return the correct
// result.  The resulting point will be normalized.
func AddNonConst(p1, p2, result *JacobianPoint) {
```

Therefore, they do need to be normalized to satisfy the API contract.

[the remark this answers is not in the document]
Comment (rstaudt2) [reply dcrec/secp256k1/curve.go:282]: That makes sense, thanks for the explanation!

[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve.go:282]: I updated the comments (as a part of the comment update commit PR this one depends on) to remove the part about a magnitude of 1 since it was misleading as pointed out.

File: dcrec/secp256k1/curve.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve.go:662]: I made the suggested change since it seems clearer to you and likely to will others as a result.

**Review by JoeGruffins — APPROVED** [review]
Code looks correct to me. Have synced from zero on mainnet and testnet with the same flags as @rstaudt2 . Have run this compiled with other projects and no problems. The explanations of the "magic" are extremely helpful.

File: dcrec/secp256k1/curve.go [hunk 6]
```diff
@@ -243,10 +275,10 @@ func addZ1EqualsZ2(p1, p2, result *JacobianPoint) {
 	e.Mul2(x1, &b)                         // E = X1*B (mag: 1)
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve.go:275]: Good catch.  Updated.
